RickAndMorty-iOS/Views/Characters/CharacterDetailsViewController.swift

```diff
@@ -0,0 +1,253 @@
+//
+//  CharacterDetailsViewController.swift
+//  RickAndMorty-iOS
+//
+//  Created by Gagan on 2023-03-10.
+//
+
+import UIKit
+import Combine
+import SDWebImage
+
+class CharacterDetailsViewController: UIViewController {
+
+    let viewModel = CharacterDetailsViewModel()
+    let locationsViewModel = LocationsViewModel()
+    var characterDetailsView = CharacterDetailsView()
+    weak var coordinator: MainCoordinator?
+    var characterID: String?
+    var avatarImageUrl: String?
+    var titleViewState: TitleViewState = .noTitle
+
+    private var dataSource: DataSource!
+    typealias DataSource = UICollectionViewDiffableDataSource<Section, AnyHashable>
+    typealias Snapshot = NSDiffableDataSourceSnapshot<Section, AnyHashable>
+    private var cancellables = Set<AnyCancellable>()
+    var snapshot = Snapshot()
+
+    enum TitleViewState {
+        case noTitle, title, titleWithImage
+    }
+
+    override func loadView() {
+        view = characterDetailsView
+        characterDetailsView.collectionView.delegate = self
+    }
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        viewModel.selectedCharacter = characterID!
+        configureDataSource()
+        showEmptyData()
+        subscribeToViewModel()
+        updateTitleView()
+    }
+
+    func showEmptyData() {
+        snapshot.deleteAllItems()
+        snapshot.appendSections([.appearance, .info, .location, .episodes, .empty, .emptyInfo, .emptyLocation])
+        snapshot.appendItems(Array(repeatingExpression: EmptyData(id: UUID()), count: 1), toSection: .empty)
+        snapshot.appendItems(Array(repeatingExpression: EmptyData(id: UUID()), count: 3), toSection: .emptyInfo)
+        snapshot.appendItems(Array(repeatingExpression: EmptyData(id: UUID()), count: 2), toSection: .emptyLocation)
+        self.dataSource.apply(snapshot, animatingDifferences: true)
+    }
+
+    func subscribeToViewModel() {
+        viewModel.character.sink(receiveValue: { [self] characterInfo in
```

Comment:
in all closures you should pass [weak self] reference. Passing strong reference could lead to memory leakage.